aggregator/apps/api/v1/api_client.py

```diff
@@ -92,11 +107,86 @@ def get_data_for_address(self, slug):
         else:
             wcivf_url = f'{settings.WCIVF_BASE_URL}candidates_for_postcode/?postcode={wdiv_json["addresses"][0]["postcode"]}'
 
-        wcivf_response = asyncio.ensure_future(fetch(wcivf_url))
-        self.loop.run_until_complete(wcivf_response)
-        wcivf_result = wcivf_response.result()
-        if wcivf_result["status"] >= 400:
-            raise ApiError(wcivf_result["status"], wcivf_result["json"]["detail"])
+        wcivf_result = proxy_single_request(self.loop, wcivf_url, self.wcivf_params)
         wcivf_json = wcivf_result["json"]
 
         return wdiv_json, wcivf_json
+
+
+class EEApiClient(AsyncApiClient):
+    def clean_query_params(self, params):
+        """
+        Take a django QueryDict object and return another QueryDict
+        containing only the specified keys.
+
+        We should use this both:
+        - before passing a query string from devs.DC to EE and
+        - before passing a URL with a query string from EE
+          back to the devs.DC user.
+
+        There are several things this acheives:
+        1. There are some params that the devs.DC API accepts
+           that we don't want to pass through to the EE API (e.g: auth_token)
+        2. There are some params supported by the EE API
+           that we don't want to expose to dev.DC users (e.g: postcode)
+        3. In future, if we wanted to pass additional params to EE
+           (e.g: a tracking param or API key, like we do with WDIV),
+           they'll come back reflected in the next/previous URLs.
+           Cleaning prevents us exposing those params to the devs.DC user.
+        """
+        allowed_query_params = [
```

Comment:
Yep. Whether you block stuff you don't want or allow stuff you do want, we have to remember to update your list under some situation.

Essentially the tradeoff is:

Allow list:
If we add a new param to the EE API which we want to expose to users, we've got to remember to add it to the allow list when we update the documentation over here.

Block list:
If we start passing anything extra between devs.DC and EE internally (e.g: an API key), we have to remember to add it to the deny list, otherwise the default position is its going to be automatically reflected to the user.

I did think about both ways, but I think the first option is safer in that respect. Also, explicitly stating what params can be passed also serves as a reference for what should be documented.

Reply:
👍